htlcswitch/switch.go

```diff
@@ -1170,6 +1194,40 @@ func (s *Switch) handlePacketForward(packet *htlcPacket) error {
 	}
 }
 
+// CheckCircularForward checks whether a forward is circular (arrives and
+// departs on the same link) and returns a link error if the switch is
+// configured to disallow this behaviour.
+func CheckCircularForward(incoming, outgoing lnwire.ShortChannelID,
+	allowCircular bool, paymentHash lntypes.Hash) *LinkError {
+
+	// Check that the incoming link is different to the outgoing link.
+	// If they are equal, the htlc is part of a circular route which may
+	// be used to lock up our liquidity.
+	isCircularRoute := incoming.ToUint64() == outgoing.ToUint64()
+
+	// If the route is not circular we do not need to perform any further
+	// checks.
+	if !isCircularRoute {
+		return nil
+	}
+
+	// If the switch is configured to allow circular routes, log that we
+	// are allowing the route then return nil.
+	if allowCircular {
+		log.Info("allowing circular route over link: %v "+
```

Comment:
Do we want to info log each time this happens? Maybe a debug log is enough.

Reply:
Wasn't expecting it to happen too often, since it's the case where somebody has changed their config and there's a strange route going through their node, but downgraded to debug. Would imagine that operators who are using these special case routes are running debug logs anyway.